tools/crm_mon_print.c

```diff
@@ -1243,42 +1238,10 @@ print_node_attributes(mon_state_t *state, pe_working_set_t *data_set, unsigned i
 static void
 print_cluster_summary_header(mon_state_t *state)
 {
-    switch (state->output_format) {
-        case mon_output_html:
-        case mon_output_cgi:
-            fprintf(state->stream, " <h2>Cluster Summary</h2>\n <p>\n");
-            break;
-
-        case mon_output_xml:
-            fprintf(state->stream, "    <summary>\n");
-            break;
-
-        default:
-            break;
-    }
-}
-
-/*!
- * \internal
- * \brief Print footer for cluster summary if needed
- *
- * \param[in] stream     File stream to display output to
- */
-static void
-print_cluster_summary_footer(mon_state_t *state)
-{
-    switch (state->output_format) {
-        case mon_output_cgi:
-        case mon_output_html:
-            fprintf(state->stream, " </p>\n");
-            break;
-
-        case mon_output_xml:
-            fprintf(state->stream, "    </summary>\n");
-            break;
-
-        default:
-            break;
+    if (state->output_format == mon_output_xml) {
+        state->out->begin_list(state->out, "summary", NULL, NULL);
+    } else {
+        state->out->begin_list(state->out, "Cluster Summary", NULL, NULL);
     }
```

Comment:
Not with the simple-list option set, which should be forced on if the output format is XML.

Reply:
Ah, I overlooked that. Just thinking out loud, since this is all just for backward compatibility and hopefully won't expand beyond this, what do you think of moving this to begin_list_xml() with e.g. if (simple_list) { if !strcmp(name, "Cluster Summary") name = "summary" else if ... }?

Reply:
Sorry, I'm not quite following you here.

Reply:
I hate when a caller needs to know the output format, since it goes against the design. In this case we need it for backward compatibility, but we shouldn't need such hacks outside of crm_mon going forward. So I'm wondering which is better -- putting the if/else in crm_mon, or moving it to xml_begin_list() like:

crm_mon:

    state->out->begin_list(state->out, "Cluster Summary", NULL, NULL);

output_xml.c:

    if (legacy_xml) {
        if (!strcmp(name, "Cluster Summary")) {
            name = "summary";
        else if ...
    }

Since crm_mon will be the only thing using legacy_xml, it shouldn't be a performance hit for any other callers. Of course it's ugly either way, it's just a question of whether the ugliness should be in crm_mon or output_xml.

Reply:
The argument for keeping it in crm_mon is so it doesn't clutter the general-purpose code, the argument for putting it in output_xml is that it's legacy-XML-specific handling.

Reply:
Oh, I see now.  There's a potential third way that accomplishes the same thing without cluttering up the general purpose code.  I could add a tools/crm_mon_xml.c (which we've successfully avoided so far) that redefines begin_list to do the name translation.

Reply:
Yeah, that seems to work pretty well.  It's basically just a new file with a big if-else block for all the possible names.  Everything is contained in that one spot.  One downside is there's no way to see if legacy_xml or simple_list are set (they're static to lib/common/output_xml.c) so crm_mon will be forced to just assume that legacy_xml is always true.  We force that in crm_mon.c so it's not really a big deal.

I think this works best as a new patch since it touches so many other patches.  Also, there's still several checks against output_format I cannot get rid of, though most are in crm_mon.c itself.